include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,671 @@
+            ::std::forward<_Range2>(__bins), __func, ::std::forward<_Range3...>(__opt_range)...));
+    }
+    else // otherwise, use global atomics (private copies per workgroup)
+    {
+        //Use __iters_per_work_item here as a runtime parameter, because only one kernel is created for
+        // private_global_atomics with a variable number of iterations per workitem. __iters_per_work_item is just a
+        // suggestion which but global memory limitations may increase this value to be able to fit the workgroup
+        // private copies of the histogram bins in global memory.  No unrolling is taken advantage of here because it
+        // is a runtime argument.
+        return __future(__histogram_general_private_global_atomics(
+            ::std::forward<_ExecutionPolicy>(__exec), __init_e, __iters_per_work_item, __work_group_size,
+            ::std::forward<_Range1>(__input), ::std::forward<_Range2>(__bins), __func,
+            ::std::forward<_Range3...>(__opt_range)...));
+    }
+}
+
+template <::std::uint16_t __iters_per_work_item, typename _ExecutionPolicy, typename _Range1, typename _Range2,
+          typename _IdxHashFunc, typename... _Range3>
+inline auto
+__parallel_histogram_impl(_ExecutionPolicy&& __exec, const sycl::event& __init_e, _Range1&& __input, _Range2&& __bins,
+                          _IdxHashFunc __func, /*req_sycl_conversion = */ ::std::false_type, _Range3&&... __opt_range)
+{
+    //wrap binhash in a wrapper to allow shared memory boost where available
+    return __parallel_histogram_select_kernel<__iters_per_work_item>(
+        ::std::forward<_ExecutionPolicy>(__exec), __init_e, ::std::forward<_Range1>(__input),
+        ::std::forward<_Range2>(__bins), __binhash_SLM_wrapper(__func), ::std::forward<_Range3...>(__opt_range)...);
+}
+
+template <::std::uint16_t __iters_per_work_item, typename _ExecutionPolicy, typename _Range1, typename _Range2,
+          typename _Range3>
+inline auto
+__parallel_histogram_impl(_ExecutionPolicy&& __exec, const sycl::event& __init_e, _Range1&& __input, _Range2&& __bins,
+                          oneapi::dpl::__internal::__custom_range_binhash<_Range3> __func,
+                          /*req_sycl_conversion = */ ::std::true_type)
+{
+    auto __range_to_upg = __func.get_range();
+    //required to have this in the call stack to keep any created buffers alive
+    auto __keep_boundaries =
+        oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::read,
+                                                decltype(__range_to_upg.begin())>();
+    auto __boundary_buf = __keep_boundaries(__range_to_upg.begin(), __range_to_upg.end());
+    auto __boundary_view = __boundary_buf.all_view();
+    auto __bin_hash = oneapi::dpl::__internal::__custom_range_binhash(__boundary_view);
+    return __parallel_histogram_impl<__iters_per_work_item>(
```

Comment:
You have "interesting" `Range3` optional parameter for the overload above, but I cannot see how it's further used except `__require_access` function

Reply:
Yes, the `Range3` sequence is used from within the "custom range" binhash structure to hold the boundary sequence (but not used for the evenly divided binhash structure).  This is why it is optional and passed along to the kernels to be required.  It is used by the binhash structure internally (if applicable).

I should be able to work out a way to minimize the number of levels of indirection this optional parameter is passed through by encoding / using some information in the binhash function closer to the kernel launch.  I'll explore that, as it would improve the readability of this code.

Reply:
I've fixed this to fold completely into the binhash structure to be required in the submit call.  In hindsight, this seems the obvious way to do it, and simplifies the code a lot, thanks.